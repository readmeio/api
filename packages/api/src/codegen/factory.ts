import type CodeGenerator from './codegenerator.js';
import type Storage from '../storage.js';
import type Oas from 'oas';

import TSGenerator from './languages/typescript/index.js';

export enum SupportedLanguages {
  JS = 'js',
}
export interface InstallerOptions {
  /**
   * Will initiate a dry run install process. Used for simulating installations within a unit test.
   */
  dryRun?: boolean;

  /**
   * Used for stubbing out the logger that we use within the installation process so it can be
   * easily introspected without having to mock out `console.*`.
   */
  logger?: (msg: string) => void;
}

export function codegenFactory(
  language: SupportedLanguages,
  spec: Oas,
  specPath: string,
  identifier: string,
): CodeGenerator {
  switch (language) {
    case SupportedLanguages.JS:
      return new TSGenerator(spec, specPath, identifier);

    default:
      throw new TypeError(`Unsupported language supplied: ${language}`);
  }
}

export function uninstallerFactory(
  language: SupportedLanguages,
  storage: Storage,
  opts: InstallerOptions = {},
): Promise<void> {
  switch (language) {
    case SupportedLanguages.JS:
      return TSGenerator.uninstall(storage, opts);

    default:
      throw new TypeError(`Unsupported language supplied: ${language}`);
  }
}
