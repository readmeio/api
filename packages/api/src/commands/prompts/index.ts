import promptTerminal from '../../lib/prompt.js';
import Storage from '../../storage.js';

export async function confirmProposedIdentifier(initial: string, proposed: string) {
  return promptTerminal({
    type: 'confirm',
    name: 'value',
    message: `Looks like this API is called "${initial}", would you like to install it as "@api/${proposed}"?`,
  });
}

export async function promptForIdentifier(sdkExplanation: boolean = true) {
  let message = 'What would you like to identify this API as?';
  if (sdkExplanation) {
    message += ' This will be how you use the SDK.';
  }

  message += ' (e.g. entering `petstore` would result in `@api/petstore`)';

  return promptTerminal({
    type: 'text',
    name: 'value',
    message,
    validate: value => {
      if (!value) {
        return false;
      }

      try {
        return Storage.isIdentifierValid(value, true);
      } catch (err) {
        return err.message;
      }
    },
  });
}
