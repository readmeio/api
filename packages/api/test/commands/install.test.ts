import type { SpyInstance } from 'vitest';

import { CommanderError } from 'commander';
import prompts from 'prompts';
import uniqueTempDir from 'unique-temp-dir';
import { describe, beforeEach, it, expect, vi, afterEach } from 'vitest';

import { SupportedLanguages } from '../../src/codegen/factory.js';
import installCmd from '../../src/commands/install.js';
import Storage from '../../src/storage.js';

const baseCommand = ['api', 'install'];

const cmdError = (msg: string) => new CommanderError(0, '', msg);

describe('install command', () => {
  let stdout: string[];
  let stderr: string[];
  let consoleLogSpy: SpyInstance;

  beforeEach(() => {
    stdout = [];
    stderr = [];

    installCmd.exitOverride();
    installCmd.configureOutput({
      writeOut: str => stdout.push(str),
      writeErr: str => stderr.push(str),
    });

    consoleLogSpy = vi.spyOn(console, 'log').mockImplementation(() => {});
    Storage.setStorageDir(uniqueTempDir());
  });

  afterEach(() => {
    consoleLogSpy.mockRestore();
    Storage.reset();
  });

  it('should error out if uri is not passed', () => {
    return expect(installCmd.parseAsync([...baseCommand])).rejects.toStrictEqual(
      cmdError("error: missing required argument 'uri'"),
    );
  });

  it('should error out if invalid uri is passed', () => {
    return expect(installCmd.parseAsync([...baseCommand, 'petstore.json'])).rejects.toThrow(
      /Sorry, we were unable to load an API definition from .*petstore.json. Please either supply a URL or a path on your filesystem./,
    );
  });

  it('should accept valid lang parameter but error out if invalid uri is passed', () => {
    return expect(
      installCmd.parseAsync([...baseCommand, 'petstore.json', '--lang', SupportedLanguages.JS]),
    ).rejects.toThrow(
      /Sorry, we were unable to load an API definition from .*petstore.json. Please either supply a URL or a path on your filesystem./,
    );
  });

  it('should error out if invalid lang is passed', () => {
    return expect(installCmd.parseAsync([...baseCommand, '--lang', 'javascript'])).rejects.toStrictEqual(
      cmdError("error: option '-l, --lang <language>' argument 'javascript' is invalid. Allowed choices are js."),
    );
  });

  it('should handle user answering no to package installation confirmation prompt', () => {
    prompts.inject(['petstore', false]);
    return expect(
      installCmd.parseAsync([...baseCommand, '../test-utils/definitions/simple.json']),
    ).rejects.toStrictEqual(new Error('Installation cancelled.'));
  });

  it('should print help screen', async () => {
    await expect(installCmd.parseAsync([...baseCommand, '--help'])).rejects.toStrictEqual(cmdError('(outputHelp)'));

    expect(stdout.join('\n')).toMatchSnapshot();
  });

  it.todo('should surface generation errors');
  it.todo('should surface file save errors');
  it.todo('should surface package installation errors');
  it.todo('should surface compilation errors');
  it.todo('should successfully generate SDK');
  it.todo('should successfully bypass all prompts with --yes option');
});
