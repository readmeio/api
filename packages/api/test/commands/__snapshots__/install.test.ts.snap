// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`install command > should print help screen 1`] = `
"Usage: install [options] <uri>

install an API SDK into your codebase

Arguments:
  uri                            an API to install

Options:
  -i, --identifier <identifier>  API identifier (eg. \`@api/petstore\`)
  -l, --lang <language>          SDK language (choices: "js", default: "js")
  -y, --yes                      Automatically answer "yes" to any prompts
                                 printed
  -h, --help                     display help for command


Examples:
  $ npx api install @developers/v2.0#nysezql0wwo236
  $ npx api install https://raw.githubusercontent.com/readmeio/oas-examples/main/3.0/json/petstore-simple.json
  $ npx api install ./petstore.json
"
`;
